crates/typst-layout/src/rules.rs

```diff
@@ -435,34 +435,12 @@ const BIBLIOGRAPHY_RULE: ShowFn<BibliographyElem> = |elem, engine, styles| {
     let span = elem.span();
 
     let mut seq = vec![];
-    if let Some(title) = elem.title.get_ref(styles).clone().unwrap_or_else(|| {
-        Some(
-            TextElem::packed(Packed::<BibliographyElem>::local_name_in(styles))
-                .spanned(span),
-        )
-    }) {
-        seq.push(
-            HeadingElem::new(title)
-                .with_depth(NonZeroUsize::ONE)
-                .pack()
-                .spanned(span),
-        );
+    if let Some(title) = elem.realize_title(styles) {
+        seq.push(title);
     }
 
-    let works = Works::generate(engine).at(span)?;
-    let references = works
-        .references
-        .as_ref()
-        .ok_or_else(|| match elem.style.get_ref(styles).source {
-            CslSource::Named(style, _) => eco_format!(
-                "CSL style \"{}\" is not suitable for bibliographies",
-                style.display_name()
-            ),
-            CslSource::Normal(..) => {
-                "CSL style is not suitable for bibliographies".into()
-            }
-        })
-        .at(span)?;
+    let works = elem.realize_works(engine, styles)?;
+    let references = works.references.as_ref().unwrap();
```

Comment:
This is a bit of an anti-pattern as it doesn't seem, at first, evident that `realize_works` would guarantee that `references` is `Some`.

The following would be more idiomatic (with the appropriate changes to `realize_works`).

```suggestion
    let (works, references) = elem.realize_works(engine, styles)?;
```